File: 6/task6.py
from collections import defaultdict
from pathlib import Path

PairType = tuple[int, int]
VisitsType = dict[PairType, set[PairType]]

field: list[list[bool]] = []  # True - прeпятствие, иначе False  # noqa: RUF003
start_position: tuple[int, int] = -1, -1
with Path(__file__).with_name('input.txt').open('r') as f:
    for i, row in enumerate(f):
        j = row.find('^')
        if j != -1:
            start_position = (i, j)
        field.append([x == '#' for x in row.strip()])

rows_count = len(field)
cols_count = len(field[0])


class InLoopError(Exception):
    pass


DIRS: list[PairType] = [(-1, 0), (0, 1), (1, 0), (0, -1)]


def traverse(field: list[list[str]]) -> VisitsType:
    visits: VisitsType = defaultdict(set)
    direction = 0
    position = start_position
    while True:
        visits[position].add(direction)
        newx, newy = map(sum, zip(position, DIRS[direction], strict=True))
        if newx not in range(rows_count) or newy not in range(cols_count):
            return visits
        if field[newx][newy]:
            direction = (direction + 1) % len(DIRS)
            continue
        if direction in visits[newx, newy]:
            raise InLoopError
        position = newx, newy


visits = traverse(field)

obstacles = 0
for cell in visits:
    field[cell[0]][cell[1]] = True
    try:
        traverse(field)
    except InLoopError:
        obstacles += 1
    field[cell[0]][cell[1]] = False

print(len(visits), obstacles)  # noqa: T201

File: 6/task6.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"slices"
)

type Coord struct {
	x int
	y int
}
type Visits = map[Coord][]Coord

var DIRS = [...]Coord{
	{-1, 0},
	{0, 1},
	{1, 0},
	{0, -1},
}

func (c Coord) move(other Coord) Coord {
	return Coord{c.x + other.x, c.y + other.y}
}

func traverse(field [][]bool, start_position Coord) (Visits, error) {
	visits := make(Visits)
	direction := 0
	position := start_position
	rows_count := len(field)
	cols_count := len(field[0])

	for {
		visits[position] = append(visits[position], DIRS[direction])
		newPos := position.move(DIRS[direction])
		if newPos.x < 0 || newPos.y < 0 || newPos.x >= rows_count || newPos.y >= cols_count {
			return visits, nil
		}
		if field[newPos.x][newPos.y] {
			direction = (direction + 1) % len(DIRS)
			continue
		}
		if slices.Contains(visits[newPos], DIRS[direction]) {
			return nil, fmt.Errorf("loop detected")
		}
		position = newPos
	}
}
func main() {
	field := [][]bool{}
	startPosition := Coord{-1, -1}

	file, err := os.Open("input.txt")
	if err != nil {
		panic(err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	x := -1
	for scanner.Scan() {
		line := scanner.Text()
		x++
		row := make([]bool, 0, len(line))
		for y, ch := range line {
			if ch == '^' {
				startPosition = Coord{x, y}
			}
			row = append(row, ch == '#')
		}
		field = append(field, row)
	}

	visits, err := traverse(field, startPosition)
	if err != nil {
		panic(err)
	}
	fmt.Printf("Result 1: %d \n", len(visits))

	obstacles := 0
	for cell := range visits {
		field[cell.x][cell.y] = true
		_, err = traverse(field, startPosition)
		if err != nil {
			obstacles += 1
		}
		field[cell.x][cell.y] = false
	}
	fmt.Printf("Result 2: %d \n", obstacles)
}
